src/coreclr/jit/assertionprop.cpp

```diff
@@ -1972,21 +1981,66 @@ AssertionIndex Compiler::optAddAssertion(AssertionDsc* newAssertion)
         return NO_ASSERTION_INDEX;
     }
 
+    AssertionIndex fastAnswer = NO_ASSERTION_INDEX;
+    AssertionIndex slowAnswer = NO_ASSERTION_INDEX;
+#ifdef DEBUG
+    bool found = false;
+
     // Check if exists already, so we can skip adding new one. Search backwards.
     for (AssertionIndex index = optAssertionCount; index >= 1; index--)
     {
         AssertionDsc* curAssertion = optGetAssertion(index);
-        if (curAssertion->Equals(newAssertion, !optLocalAssertionProp))
+        if (curAssertion->Equals(*newAssertion))
         {
-            return index;
+            found      = true;
+            slowAnswer = index;
+            break;
         }
     }
+#endif
 
     // Check if we are within max count.
     if (optAssertionCount >= optMaxAssertionCount)
     {
-        return NO_ASSERTION_INDEX;
+        // If we don't want to add more assertions, just see if `newAssertion` already
+        // exists, and return the corresponding assertionIndex.
+        if (optAssertionDscMap->Lookup(*newAssertion, &fastAnswer))
+        {
+            assert(slowAnswer == fastAnswer);
+            return fastAnswer;
+        }
+        else
+        {
+#ifdef DEBUG
+            if (found)
+            {
+                JITDUMP("HashCode= %u not found in map.\n", AssertionDscKeyFuncs::GetHashCode(*newAssertion));
+                assert(false);
+            }
+#endif
+            return NO_ASSERTION_INDEX;
+        }
+    }
+
+    // Check if newAssertion already exists and return the corresponding assertionIndex
+    // Otherwise, set it in the map.
+    if (optAssertionDscMap->Set(*newAssertion, optAssertionCount + 1, AssertionDscMap::SetKind::SkipIfExist,
+                                &fastAnswer))
```

Comment:
It is not obvious - why do we need `SetKind::SkipIfExist` if we have `LookupPointer`?

Reply:
Because you will call the `GetHashCode()` twice, first through `LookupPointer` and then during `Set`. With `SkipIfExist`, you just call it once.